hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/upgrade/BasicUpgradeFinalizer.java

```diff
@@ -0,0 +1,300 @@
+ */
+@SuppressWarnings("checkstyle:VisibilityModifier")
+public class BasicUpgradeFinalizer<T, V extends AbstractLayoutVersionManager>
+    implements UpgradeFinalizer<T> {
+
+  protected V versionManager;
+  protected String clientID;
+  protected T component;
+
+  private Queue<String> msgs = new ConcurrentLinkedQueue<>();
+  protected boolean isDone = false;
+
+  public BasicUpgradeFinalizer(V versionManager) {
+    this.versionManager = versionManager;
+  }
+
+  public StatusAndMessages preFinalize(String upgradeClientID, T id)
+      throws UpgradeException {
+    switch (versionManager.getUpgradeState()) {
+    case STARTING_FINALIZATION:
+      return STARTING_MSG;
+    case FINALIZATION_IN_PROGRESS:
+      return FINALIZATION_IN_PROGRESS_MSG;
+    case FINALIZATION_DONE:
+    case ALREADY_FINALIZED:
+      return FINALIZED_MSG;
+    default:
+      if (!versionManager.needsFinalization()) {
+        throw new UpgradeException("Upgrade found in inconsistent state. " +
+            "Upgrade state is FINALIZATION_REQUIRED while MLV has been " +
+            "upgraded to SLV.", INVALID_REQUEST);
+      }
+      versionManager.setUpgradeState(STARTING_FINALIZATION);
+
+      clientID = upgradeClientID;
+      this.component = id;
+      return FINALIZATION_REQUIRED_MSG;
+    }
+  }
+
+  /*
+   * This method must be overriden by the component implementing the
+   * finalization logic.
+   */
+  public StatusAndMessages finalize(String upgradeClientID, T id)
+      throws IOException {
+    StatusAndMessages response = preFinalize(upgradeClientID, id);
+    if (response.status() != FINALIZATION_REQUIRED) {
+      return response;
+    }
+
+    /**
+     * Overriding class should schedule actual finalization logic
+     * in a separate thread here.
+     */
+    return STARTING_MSG;
+  }
+
+  @Override
+  public synchronized StatusAndMessages reportStatus(
+      String upgradeClientID, boolean takeover
+  ) throws UpgradeException {
+    if (takeover) {
+      clientID = upgradeClientID;
+    }
+    assertClientId(upgradeClientID);
+    List<String> returningMsgs = new ArrayList<>(msgs.size()+10);
+    Status status = isDone ? FINALIZATION_DONE : FINALIZATION_IN_PROGRESS;
+    while (msgs.size() > 0) {
+      returningMsgs.add(msgs.poll());
+    }
+    return new StatusAndMessages(status, returningMsgs);
+  }
+
+  private void assertClientId(String id) throws UpgradeException {
+    if (!this.clientID.equals(id)) {
+      throw new UpgradeException("Unknown client tries to get finalization " +
+          "status.\n The requestor is not the initiating client of the " +
+          "finalization, if you want to take over, and get unsent status " +
+          "messages, check -takeover option.", INVALID_REQUEST);
+    }
+  }
+
+  protected void finalizeFeature(LayoutFeature feature, Storage config)
+      throws UpgradeException {
+    Optional<? extends UpgradeAction> action = feature.onFinalizeAction();
+
+    if (action.isPresent()) {
```

Comment:
Aren't we want to do a noop when the action is not present?

Reply:
why do we want to do noop ? action is not present we do not have to do anything.

Reply:
Sorry for this, but my wording was not correct as it seems.
Let me put it this way:
Here if action is present in the optional, we emit a noop message and return, I believe we want to emit a noop message and return if action.isPresent yields is false, so we miss a negation in the condition.

Reply:
yes, I see now :)